# Netflix Content Analysis
**Analyst:** Kushagra Yadav

**Date:** July 2026

**Dataset:** 8,807 Netflix titles | 2008-2021

## Objective
To analyze Netflix's content library and uncover key insights around content type distribution, country-wise production, genre trends, and yearly growth - with actionable recommendations for content strategy.

## Tools Used
Python (Pandas, Matplotlib, Seaborn) | SQL (SQLite) | Power BI | Excel | Git

---
## Importing Libraries and Loading dataset

In [23]:
import pandas as pd
import numpy as np

df = pd.read_excel("netflix_titles.xlsx")

print("Shape:", df.shape)
print("\nColumns:",list(df.columns))
print("\nData Types:")
print(df.dtypes)
df.head()

Shape: (8807, 12)

Columns: ['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added', 'release_year', 'rating', 'duration', 'listed_in', 'description']

Data Types:
show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


---
### Checking for the missing values

In [25]:
print("Missing values:")
print(df.isnull().sum())

#missing value percentage
print("\nMissing percentage:")
print(round(df.isnull().sum()/len(df)*100, 2))

Missing values:
show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

Missing percentage:
show_id          0.00
type             0.00
title            0.00
director        29.91
cast             9.37
country          9.44
date_added       0.11
release_year     0.00
rating           0.05
duration         0.03
listed_in        0.00
description      0.00
dtype: float64


---
## Checking the content type breakdown

In [27]:
print("Content Type Breakdown:")
print(df['type'].value_counts())

print("\nPercentage:")
print(round(df['type'].value_counts()/len(df)*100, 2))

Content Type Breakdown:
type
Movie      6131
TV Show    2676
Name: count, dtype: int64

Percentage:
type
Movie      69.62
TV Show    30.38
Name: count, dtype: float64


---
## Checking year range

In [29]:
print("Year Range:")
print("Oldest content:", df['release_year'].min())
print("Newest content:", df['release_year'].max())

print("\nContent added to netflix (date_added sample):")
print(df['date_added'].dropna().head(5))

Year Range:
Oldest content: 1925
Newest content: 2021

Content added to netflix (date_added sample):
0    September 25, 2021
1    September 24, 2021
2    September 24, 2021
3    September 24, 2021
4    September 24, 2021
Name: date_added, dtype: object


---

## Business Questions this Project will Answer

1. Which country produces the most Netflix content?
2. What is the ratio of Movies vs TV Shows on Netflix?
3. Which genres dominate the Netflix library?
4. How has Netflix grown its content library year over year - especially during COVID (2019-2021)?
5. What content ratings are most common and what does that say about Netflix's target audience?

---
**1. Why did i choose "fill" for some and "drop rows" for others?**

Some columns(director) is not important because **director** -> 29.91% missing value -> fill "Unknown" -> not used in analysis but kept for completeness.
Some columns(cast, country, rating, duration, date_added) are very important for the analysis and these columns are important because:
- **cast** -> 9% missing -> fill "Unknown" -> not directly used in business questions.
- **country** -> 9% missing -> fill "Unknown" -> *CRITICAL* — needed for top countries.
- **date_added** -> 0.11% missing -> drop those 10 rows -> tiny amount, won't affect analysis.
- **rating** -> 0.05% missing -> drop those 4 rows -> tiny amount, won't affect analysis.
- **duration** -> 0.03% missing -> drop those 3 rows -> tiny amount, won't affect analysis.


**2. For each null column - why did i choose fill vs drop? What percentage threshold makes dropping rows acceptable vs unacceptable?**

**NULL HANDLING STRATEGY**
- Rule 1: If null % is above 5% AND column is important -> fill with "Unknown"
- Rule 2: If null % is below 1% -> drop those rows safely (minimal data loss)
- Rule 3: If column is not used in any business question -> fill "Unknown" and ignore

**Applied to this dataset:**
- director  -> 29.91% missing -> fill "Unknown" (not used in analysis)
- cast      -> 9.37%  missing -> fill "Unknown" (not used in analysis)
- country   -> 9.44%  missing -> fill "Unknown" (used in Q1 - handle carefully)
- date_added-> 0.11%  missing -> drop 10 rows  (safe - only 10 rows lost)
- rating    -> 0.05%  missing -> drop 4 rows   (safe - only 4 rows lost)
- duration  -> 0.03%  missing -> drop 3 rows   (safe - only 3 rows lost)
---

# Data Cleaning

In [3]:
import pandas as pd
import numpy as np

df = pd.read_excel("netflix_titles.xlsx")
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


---
## Structure Check 

In [5]:
print("Shape:", df.shape)
print("\n Data Types:")
print(df.dtypes)

Shape: (8807, 12)

 Data Types:
show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object


---
## Duplicates checking

In [7]:
print("Duplicate rows:", df.duplicated().sum())
print("\nDuplicate titles:", df['title'].duplicated().sum())

Duplicate rows: 0

Duplicate titles: 3


In [9]:
df = df.drop_duplicates()
print("Shape after removing duplicates:", df.shape)

Shape after removing duplicates: (8807, 12)


**Duplicate check result:** 0 fully duplicate rows found. 3 titles appear more than once - likely the same title released in different years/countries (e.g. a remake), not true duplicates. No rows dropped.

---
## Handling nulls in high-missing columns

In [33]:
# Filling high-missing columns with Unknown [director, cast, country]
df['director'] = df['director'].fillna('Unknown')
df['cast'] = df['cast'].fillna('Unknown')
df['country'] = df['country'].fillna('Unknown')

print("Nulls after filling:")
print(df[['director','cast','country']].isnull().sum())

Nulls after filling:
director    0
cast        0
country     0
dtype: int64


In [14]:
# Dropping rows with tiny null counts
df = df.dropna(subset = ['date_added', 'rating', 'duration'])

print("Shape after dropping tiny nulls:", df.shape)
print("\nRemaining nulls:")
print(df.isnull().sum())

Shape after dropping tiny nulls: (8790, 12)

Remaining nulls:
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


---
## Fixing columns 

In [17]:
print("Current data type of date_added:", df['date_added'].dtype)


df['date_added'] = pd.to_datetime(df['date_added'], errors = 'coerce')

df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month

print("Date Conversion done.")
print(df[['date_added','year_added','month_added']].head(5))

Current data type of date_added: object
Date Conversion done.
  date_added  year_added  month_added
0 2021-09-25      2021.0          9.0
1 2021-09-24      2021.0          9.0
2 2021-09-24      2021.0          9.0
3 2021-09-24      2021.0          9.0
4 2021-09-24      2021.0          9.0


In [19]:
df.head(5)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Unknown,United States,2021-09-25,2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm...",2021.0,9.0
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",2021.0,9.0
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,2021.0,9.0
3,s4,TV Show,Jailbirds New Orleans,Unknown,Unknown,Unknown,2021-09-24,2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",2021.0,9.0
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,2021.0,9.0


In [21]:
# Fixing duration column

df['duration_int'] = df['duration'].str.extract(r'(\d+)').astype(float)
df['duration_type'] = df['duration'].str.extract(r'([A-Za-z]+)')

print("Duration examples:")
print(df[['type','duration','duration_int','duration_type']].head(10))

Duration examples:
      type   duration  duration_int duration_type
0    Movie     90 min          90.0           min
1  TV Show  2 Seasons           2.0       Seasons
2  TV Show   1 Season           1.0        Season
3  TV Show   1 Season           1.0        Season
4  TV Show  2 Seasons           2.0       Seasons
5  TV Show   1 Season           1.0        Season
6    Movie     91 min          91.0           min
7    Movie    125 min         125.0           min
8  TV Show  9 Seasons           9.0       Seasons
9    Movie    104 min         104.0           min


---
## Explode genres column

In [35]:
# Splitting genres and exploding into separate rows
df['genre'] = df['listed_in'].str.split(', ')
df_genres = df.explode('genre')

print("Original shape:", df.shape)
print("After exploding genres:", df_genres.shape)
print("\nTop 10 genres:")
print(df_genres['genre'].value_counts().head(10))

Original shape: (8790, 17)
After exploding genres: (19294, 17)

Top 10 genres:
genre
International Movies        2752
Dramas                      2426
Comedies                    1674
International TV Shows      1349
Documentaries                869
Action & Adventure           859
TV Dramas                    762
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64


---
## Checking for unique values

In [27]:
print("Content types:", df['type'].unique())
print("\nRatings:", df['rating'].unique())
print("\nTop 10 countries:")
top_countries = df[df['country'] != 'Unknown']['country'].value_counts().head(10)
print(top_countries)
print("\nYear added range:", df['year_added'].min(), "to", df['year_added'].max())

Content types: ['Movie' 'TV Show']

Ratings: ['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' 'NR' 'TV-Y7-FV' 'UR']

Top 10 countries:
country
United States     2809
India              972
United Kingdom     418
Japan              243
South Korea        199
Canada             181
Spain              145
France             124
Mexico             110
Egypt              106
Name: count, dtype: int64

Year added range: 2008.0 to 2021.0


**Note on country analysis:** 831 titles (9.4%) have missing country data, filled as "Unknown" earlier for dataset completeness. When ranking countries specifically for Business Question, these are excluded so a placeholder value never appears as a false "top country."

In [30]:
multi_country = df[df['country'].str.contains(',', na=False)]
print("Titles with multiple countries listed:", len(multi_country))
print(multi_country['country'].head(10))

Titles with multiple countries listed: 1320
7      United States, Ghana, Burkina Faso, United Kin...
12                               Germany, Czech Republic
29                          United States, India, France
38                          China, Canada, United States
46                    South Africa, United States, Japan
90                                  Spain, United States
94                         United Kingdom, United States
95                         United States, United Kingdom
99                                 France, United States
111                             United States, Singapore
Name: country, dtype: object


In [42]:
# Splitting and exploding country column (same technique as genre)
df['country_list'] = df['country'].str.split(', ')
df_countries = df.explode('country_list')

print("Original shape:", df.shape)
print("After exploding countries:", df_countries.shape)

Original shape: (8790, 18)
After exploding countries: (10828, 18)


*Use **df_countries** only for country-level analysis - other business questions should use the original **df** to avoid double-counting titles*

In [44]:
print("Top 10 countries (exploded, excluding Unknown):")
top_countries_real = df_countries[df_countries['country_list'] != 'Unknown']['country_list'].value_counts().head(10)
print(top_countries_real)

Top 10 countries (exploded, excluding Unknown):
country_list
United States     3680
India             1046
United Kingdom     803
Canada             445
France             393
Japan              316
Spain              232
South Korea        231
Germany            226
Mexico             169
Name: count, dtype: int64


**Note on multi-country titles:** 1,320 titles (15%) list more than one country (co-productions). These were split and exploded so each contributing country gets proper credit, instead of being grouped under one combined string like "United States, India, France."

---
## Saving cleaned dataset

In [49]:
df.to_csv("Netflix_Cleaned.csv", index = False)
print("Cleaned file saved!")
print("Final shape:", df.shape)

Cleaned file saved!
Final shape: (8790, 18)


## Cleaning Summary
| Column     | Issue                    | Fix Applied                 |
|------------|--------------------------|-----------------------------|
| director   | 29.91% missing           | Filled with "Unknown"       |
| cast       | 9.37% missing            | Filled with "Unknown"       |
| country    | 9.44% missing            | Filled with "Unknown"       |
| date_added | 0.11% missing (10 rows)  | Dropped rows                |
| rating     | 0.05% missing (4 rows)   | Dropped rows                |
| duration   | 0.03% missing (3 rows)   | Dropped rows                |
| date_added | Stored as text           | Converted to datetime       |
| duration   | Mixed text + numbers     | Extracted numeric value     |
| listed_in  | Multiple genres per cell | Exploded into separate rows |

**Final clean dataset: Netflix_Cleaned.csv**